src/Avalonia.Controls/Application.cs

```diff
@@ -171,27 +209,95 @@ public virtual void Initialize()
         /// <param name="closable">The closable to track</param>
         public void Run(ICloseable closable)
         {
-            var source = new CancellationTokenSource();
-            closable.Closed += OnExiting;
-            closable.Closed += (s, e) => source.Cancel();
-            Dispatcher.UIThread.MainLoop(source.Token);
+            if (_mainLoopCancellationTokenSource != null)
+            {
+                throw new Exception("Run should only called once");
+            }
+
+            closable.Closed += (s, e) => Exit();
+
+            _mainLoopCancellationTokenSource = new CancellationTokenSource();
+
+            Dispatcher.UIThread.MainLoop(_mainLoopCancellationTokenSource.Token);
+
+            if (!IsExiting)
+            {
+                OnExit?.Invoke(this, EventArgs.Empty);
+            }
+        }
+
+        /// <summary>
+        /// Runs the application's main loop until some condition occurs that is specified by ExitMode.
+        /// </summary>
+        /// <param name="mainWindow">The main window</param>
+        public void Run(Window mainWindow)
+        {
+            if (_mainLoopCancellationTokenSource != null)
+            {
+                throw new Exception("Run should only called once");
+            }
+
+            _mainLoopCancellationTokenSource = new CancellationTokenSource();
+
+            Dispatcher.UIThread.InvokeAsync(
+                () =>
+                    {
+                        if (mainWindow == null)
+                        {
+                            return;
+                        }
+
+                        if (MainWindow != null)
+                        {
+                            return;
+                        }
+
+                        if (!mainWindow.IsVisible)
+                        {
+                            mainWindow.Show();
+                        }
+
+                        MainWindow = mainWindow;
+                    },
+                DispatcherPriority.Send);
+
+            Dispatcher.UIThread.MainLoop(_mainLoopCancellationTokenSource.Token);
+
+            if (!IsExiting)
+            {
+                OnExit?.Invoke(this, EventArgs.Empty);
+            }
         }
-        
+
         /// <summary>
-        /// Runs the application's main loop until the <see cref="CancellationToken"/> is cancelled.
+        /// Runs the application's main loop until the <see cref="CancellationToken"/> is canceled.
         /// </summary>
         /// <param name="token">The token to track</param>
         public void Run(CancellationToken token)
         {
             Dispatcher.UIThread.MainLoop(token);
+
+            if (!IsExiting)
+            {
+                OnExit?.Invoke(this, EventArgs.Empty);
+            }       
         }
 
         /// <summary>
         /// Exits the application
         /// </summary>
         public void Exit()
         {
+            IsExiting = true;
+
+            while (Windows.Count > 0)
```

Comment:
That's wrong because this line always removes the first window of the collection until none is left. A close removes the window from the list. I will add unit tests to prove that.

Reply:
@Gillibald Good